File: RhinoCommonExamples/GetScaleXform.vb
Imports System
Imports Rhino.Geometry

Namespace examples_cs
  Friend Class GetScaleXform
	  Inherits Rhino.Input.Custom.GetTransform

	Public Property Plane() As Plane
	Public Property RefPoint() As Point3d
	Public Property Scale() As Double

	Public Sub New()
	  Plane = Plane.WorldXY
	  RefPoint = Point3d.Origin
	  Scale = 1
	End Sub

	Protected Overrides Sub OnDynamicDraw(ByVal e As Rhino.Input.Custom.GetPointDrawEventArgs)
	  Dim basePoint As Point3d = Nothing
	  If TryGetBasePoint(basePoint) Then
		e.Display.DrawLine(basePoint, RefPoint, System.Drawing.Color.Black)
		e.Display.DrawPoint(RefPoint, System.Drawing.Color.Black)
		MyBase.OnDynamicDraw(e)
	  End If
	End Sub

	Public Function CalculateTransform(ByVal viewport As Rhino.Display.RhinoViewport, ByVal d As Double) As Transform
	  Dim basePoint As Point3d = Nothing
	  If Not TryGetBasePoint(basePoint) Then
		Return Transform.Identity
	  End If
'INSTANT VB NOTE: The variable plane was renamed since Visual Basic does not handle local variables named the same as class members well:
	  Dim plane_Renamed As Plane = viewport.ConstructionPlane()
	  plane_Renamed.Origin = basePoint
	  Dim v As Vector3d = RefPoint - basePoint

	  Dim len1 As Double = v.Length
	  If Math.Abs(len1) < 0.000001 Then
		Return Transform.Identity
	  End If

	  v.Unitize()
	  v = v * d
	  Dim len2 As Double = v.Length
	  If Math.Abs(len2) < 0.000001 Then
		Return Transform.Identity
	  End If

	  Scale = len2 / len1
	  Return Transform.Scale(plane_Renamed, Scale, Scale, Scale)
	End Function
	Public Overrides Function CalculateTransform(ByVal viewport As Rhino.Display.RhinoViewport, ByVal point As Point3d) As Transform
	  Dim basePoint As Point3d = Nothing
	  If Not TryGetBasePoint(basePoint) Then
		Return Transform.Identity
	  End If
	  Dim len2 As Double = (point - basePoint).Length
	  Dim len1 As Double = (RefPoint - basePoint).Length
	  If Math.Abs(len1) < 0.000001 OrElse Math.Abs(len2) < 0.000001 Then
		Return Transform.Identity
	  End If

	  Scale = len2 / len1

'INSTANT VB NOTE: The variable plane was renamed since Visual Basic does not handle local variables named the same as class members well:
	  Dim plane_Renamed As Plane = viewport.ConstructionPlane()
	  plane_Renamed.Origin = basePoint

	  Return Transform.Scale(plane_Renamed, Scale, Scale, Scale)
	End Function
  End Class
End Namespace

File: RhinoCommonExamples/GetScaleXform.cs
using System;
using Rhino.Geometry;

namespace examples_cs
{
  class GetScaleXform : Rhino.Input.Custom.GetTransform
  {
    public Plane Plane { get; set; }
    public Point3d RefPoint { get; set; }
    public double Scale { get; set; }

    public GetScaleXform()
    {
      Plane = Plane.WorldXY;
      RefPoint = Point3d.Origin;
      Scale = 1;
    }

    protected override void OnDynamicDraw(Rhino.Input.Custom.GetPointDrawEventArgs e)
    {
      Point3d basePoint;
      if (TryGetBasePoint(out basePoint))
      {
        e.Display.DrawLine(basePoint, RefPoint, System.Drawing.Color.Black);
        e.Display.DrawPoint(RefPoint, System.Drawing.Color.Black);
        base.OnDynamicDraw(e);
      }
    }

    public Transform CalculateTransform( Rhino.Display.RhinoViewport viewport, double d )
    {
      Point3d basePoint;
      if (!TryGetBasePoint(out basePoint))
        return Transform.Identity;
      Plane plane = viewport.ConstructionPlane();
      plane.Origin = basePoint;
      Vector3d v = RefPoint - basePoint;

      double len1 = v.Length;
      if (Math.Abs(len1) < 0.000001)
        return Transform.Identity;

      v.Unitize();
      v = v * d;
      double len2 = v.Length;
      if (Math.Abs(len2) < 0.000001)
        return Transform.Identity;

      Scale = len2 / len1;
      return Transform.Scale(plane, Scale, Scale, Scale);
    }
    public override Transform CalculateTransform(Rhino.Display.RhinoViewport viewport, Point3d point)
    {
      Point3d basePoint;
      if (!TryGetBasePoint(out basePoint))
        return Transform.Identity;
      double len2 = (point - basePoint).Length;
      double len1 = (RefPoint - basePoint).Length;
      if (Math.Abs(len1) < 0.000001 || Math.Abs(len2) < 0.000001)
        return Transform.Identity;

      Scale = len2 / len1;

      Plane plane = viewport.ConstructionPlane();
      plane.Origin = basePoint;

      return Transform.Scale(plane, Scale, Scale, Scale);
    }
  }
}

